### PR TITLE
Subsample pyro.param inside plate if event_dim is given

### LINKED ISSUE #238: Params inside iarange
the current behavior of params created inside an `iarange` (aka vectorized map_data) is a bit nonsensical: a vector of params of length `batch_size` will be created and used on every iteration. this means that a param gets reused for whatever data point happens to be selected at that index in a batch. this comes up eg when there is a mean-field model for a local variable.

(there is a reasonable work-around, by creating the vector of needed params globally, and only selecting from it in the iarange.)

a better behavior would be to automatically associate params to their index into the data tensor, pulling out the appropriate params for a given batch (and creating them as needed). it might be tricky to get this working, and especially to do it efficiently.

i'm flagging this as a low priority bug for now. we can think through a solution after launch.

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #238 

This adds a new optional kwarg `event_dim` to `pyro.param` statements. If `event_dim` is not specified, previous behavior is preserved i.e. all dims are considered event dims.

When `event_dim` is specified (e.g. 0, 1, 2), dims to the left are considered batch dims. When a param statement occurs inside a `pyro.plate`, the `event_dim` argument gives Pyro enough information to subsample the parameters (via `.index_select(plate.dim - event_dim, plate.ind)`).

This PR also anticipates named dimensions (possibly coming in PyTorch 1.2). By specifying `event_dim`, Pyro will have enough information to annotate plate names to the dims of a param tensor.

## Tested
- [x] tests of shape in test_valid_models.py
- [x] updated test_gradients.py to use this slick new syntax
- [x] add test of shape error

### CODE REVIEW

File: pyro/poutine/subsample_messenger.py [hunk 1]
```diff
@@ -117,7 +117,7 @@ def _subsample(name, size=None, subsample_size=None, subsample=None, use_cuda=No
         return size, subsample_size, subsample
 
     def _reset(self):
-        self.subsample = None
+        self._indices = None
```
Comment (fritzo) [inline pyro/poutine/subsample_messenger.py]: This was apparently a typo. I'm not sure when this is used, however.

File: pyro/poutine/subsample_messenger.py [hunk 2]
```diff
@@ -128,3 +129,20 @@ def _process_message(self, msg):
                 with ignore_jit_warnings():
                     msg["scale"] = torch.tensor(msg["scale"])
         msg["scale"] = msg["scale"] * self.size / self.subsample_size
+
+    def _postprocess_message(self, msg):
+        if msg["type"] == "param":
+            event_dim = msg["kwargs"].get("event_dim")
+            if event_dim is not None:
+                dim = self.dim - event_dim
```
Comment (eb8680) [inline pyro/poutine/subsample_messenger.py]: You'll need to check `self.dim is not None` to handle the case of sequential plates

File: tests/infer/test_valid_models.py [hunk 3]
```diff
@@ -215,6 +215,62 @@ def guide():
...
+    assert_ok(model, guide, Elbo())
+
+
+@pytest.mark.parametrize("subsample_size", [None, 5], ids=["full", "subsample"])
+@pytest.mark.parametrize("Elbo", [Trace_ELBO, TraceGraph_ELBO, TraceEnum_ELBO])
+@pytest.mark.parametrize("shape,ok", [
+    ((), True),
+    ((1,), True),
+    ((10,), True),
+    ((3, 1), True),
```
Comment (neerajprad) [inline tests/infer/test_valid_models.py]: The logic for param shapes looks very consistent, but just for my understanding, is there a use case that you have in mind for `(k, 1)` shaped params inside an enclosing plate?

Comment (fritzo) [reply tests/infer/test_valid_models.py]: The `k,1` might happen if there is another plate. I guess we could error if not all batch dims are accounted for.
